rp2040-hal/src/adc.rs

```diff
@@ -153,3 +154,210 @@ where
+    }
+
+    /// Check if there was a fifo overrun
+    ///
+    /// An overrun happens when the fifo is filled up faster than `read` is called to consume it.
+    ///
+    /// This function also clears the `over` bit if it was set.
+    pub fn is_over(&mut self) -> bool {
+        let over = self.adc.device.fcs.read().over().bit();
+        if over {
+            self.adc.device.fcs.modify(|_, w| w.over().set_bit());
+        }
+        over
+    }
+
+    /// Check if there was a fifo underrun
+    ///
+    /// An underrun happens when `read` is called on an empty fifo (`len() == 0`).
+    ///
+    /// This function also clears the `under` bit if it was set.
+    pub fn is_under(&mut self) -> bool {
+        let under = self.adc.device.fcs.read().under().bit();
+        if under {
+            self.adc.device.fcs.modify(|_, w| w.under().set_bit());
+        }
+        under
+    }
+
+    /// Read a single value from the fifo
+    pub fn read(&mut self) -> u16 {
+        self.adc.device.fifo.read().val().bits()
+    }
+
+    /// Stop capturing in free running mode.
+    ///
+    /// Disables the ADC fifo and stops ADC capture.
+    /// Returns the underlying `Adc` instance.
+    pub fn stop(self) -> &'a mut Adc {
+        self.adc.device.cs.modify(|_, w| w.start_many().clear_bit());
+        self.adc.device.fcs.modify(|_, w| w.en().clear_bit());
+        self.adc
+    }
+}
+
+pub struct RoundRobin(u8);
+
+impl<PIN: Channel<Adc, ID = u8>> From<PIN> for RoundRobin {
+    fn from(_: PIN) -> Self {
+        Self(1 << PIN::channel())
+    }
+}
+
+impl<A, B> From<(&mut A, &mut B)> for RoundRobin
+where
+    A: Channel<Adc, ID = u8>,
+    B: Channel<Adc, ID = u8>,
+{
+    fn from(_: (&mut A, &mut B)) -> Self {
+        Self(1 << A::channel() | 1 << B::channel())
+    }
+}
+
+impl<A, B, C> From<(&mut A, &mut B, &mut C)> for RoundRobin
+where
+    A: Channel<Adc, ID = u8>,
+    B: Channel<Adc, ID = u8>,
+    C: Channel<Adc, ID = u8>,
+{
+    fn from(_: (&mut A, &mut B, &mut C)) -> Self {
+        Self(1 << A::channel() | 1 << B::channel() | 1 << C::channel())
+    }
+}
+
+impl<A, B, C, D> From<(&mut A, &mut B, &mut C, &mut D)> for RoundRobin
+where
+    A: Channel<Adc, ID = u8>,
+    B: Channel<Adc, ID = u8>,
+    C: Channel<Adc, ID = u8>,
+    D: Channel<Adc, ID = u8>,
+{
+    fn from(_: (&mut A, &mut B, &mut C, &mut D)) -> Self {
+        Self(1 << A::channel() | 1 << B::channel() | 1 << C::channel() | 1 << D::channel())
+    }
+}
+
+impl<A, B, C, D, E> From<(&mut A, &mut B, &mut C, &mut D, &mut E)> for RoundRobin
+where
+    A: Channel<Adc, ID = u8>,
+    B: Channel<Adc, ID = u8>,
+    C: Channel<Adc, ID = u8>,
+    D: Channel<Adc, ID = u8>,
+    E: Channel<Adc, ID = u8>,
+{
+    fn from(_: (&mut A, &mut B, &mut C, &mut D, &mut E)) -> Self {
+        Self(1 << A::channel() | 1 << B::channel() | 1 << C::channel() | 1 << D::channel() | 1 << E::channel())
+    }
+}
+
+impl crate::dma::ReadTarget for Fifo<'_, true> {
+    type ReceivedWord = u8;
```

Comment:
From the datasheet: "If the DMA transfer size is set to 8 bits, so that the DMA transfers to a byte array in memory, FCS.SHIFT must also be set, to pre-shift the FIFO samples to 8 bits of significance."

Reply:
Good point, this would not work properly then (or capture the wrong values). I don't know if it's possible to use DMA transfers with 12 bit precision (seems a shame to throw away those bits). In that case we would need two different `ReadTarget` implementations (one with `ReceivedWord = u8`, one with `u16`), depending on if the user decided to FCS.SHIFT or not.